src/main/java/athena/TimeAllocator.java

```diff
@@ -53,60 +58,153 @@ public void runAllocate() throws TaskNotFoundException {
+     *
+     * @param pos                staring position
+     * @param end                ending position
+     * @param undefinedTimeTasks Tasks with undefined times
+     * @param carryOverTasks     Tasks that did not get assigned yet
+     * @return
+     */
+    private ArrayList<Integer> getBestLog(int pos, int end, ArrayList<Task> undefinedTimeTasks,
+                                          ArrayList<Task> carryOverTasks) throws AllocationFailedException {
+        ArrayList<Integer> currentLog = new ArrayList<Integer>();
+        ArrayList<Integer> bestLog = new ArrayList<Integer>();
+        boolean active = true;
+        while (active) {
+            currentLog = getLog(undefinedTimeTasks, pos, end);
+            active = currentLog.contains(-1);
+            if ((bestLog.indexOf(-1) < currentLog.indexOf(-1)) | !active) {
+                bestLog = currentLog;
+            }
+            if (!undefinedTimeTasks.isEmpty()) {
+                carryOverTasks.add(undefinedTimeTasks.get(0));
+                undefinedTimeTasks.remove(0);
+
+            } else {
+                break;
+            }
+
+        }
+        for (int taskNumber : bestLog) {
+            try {
+                Task taskRemoved = taskList.getTaskFromNumber(taskNumber);
+                undefinedTimeTasks.remove(taskRemoved);
+                carryOverTasks.remove(taskRemoved);
+            } catch (TaskNotFoundException e) {
+                throw new AllocationFailedException();
+            }
+        }
+        carryOverTasks.addAll(undefinedTimeTasks);
+        return bestLog;
+    }
+
+    /**
+     * Gets a filled vacant slots given undefinedTimeTasks.
+     *
+     * @param undefinedTimeTasks Tasks to be slotted into the timetable
+     * @param pos                starting position
+     * @param end                ending position
+     * @return ArrayList containing taskNumbers
+     */
+    private ArrayList<Integer> getLog(ArrayList<Task> undefinedTimeTasks, int pos, int end) {
+        int space = end - pos;
+        ArrayList<Integer> log = new ArrayList<Integer>(Collections.nCopies(space, -1));
+        for (Task currTask : undefinedTimeTasks) {
+            int span = currTask.getTimeInfo().getDuration();
+            if (span <= space) {
+                int taskNumber = currTask.getNumber();
+                for (int i = 0; i < span; i++) {
+                    int relativePos = end - space - pos;
+                    log.set(i + relativePos, currTask.getNumber());
                 }
+                space -= span;
+            }
+            if (space == 0) {
+                break;
+            }
+        }
+        return log;
+    }
+
 
+    /**
+     * Finds end of vacant time slot.
+     *
+     * @param dayLog log of the day's tasks
+     * @param end    ending position
+     * @param sleep  sleep time
+     * @return valid end value
+     */
+    private int getEnd(ArrayList<Integer> dayLog, int end, int sleep) {
```

Comment:
How about `nextVacantSlotEnd` or something that indicates that this finds the end of the next vacant time slot?